pkg/export/secrets/kubernetes_test.go

```diff
@@ -0,0 +1,548 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func toSecretConfig(s *corev1.Secret, key string) *KubernetesSecretConfig {
+	return &KubernetesSecretConfig{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+		Key:       key,
+	}
+}
+
+func deleteKey(s *corev1.Secret, key string) {
+	delete(s.Data, key)
+	delete(s.StringData, key)
+}
+
+func updateKey(s *corev1.Secret, key, value string) {
+	if _, ok := s.Data[key]; ok {
+		s.Data[key] = []byte(value)
+		return
+	}
+	if _, ok := s.StringData[key]; ok {
+		s.StringData[key] = value
+		return
+	}
+	panic(fmt.Errorf("invalid key %q in secret: %s/%s", key, s.Namespace, s.Name))
+}
+
+func copyKey(s *corev1.Secret, keyFrom, keyTo string) {
+	if _, ok := s.Data[keyFrom]; ok {
+		s.Data[keyTo] = s.Data[keyFrom]
+		return
+	}
+	if _, ok := s.StringData[keyFrom]; ok {
+		s.StringData[keyTo] = s.StringData[keyFrom]
+		return
+	}
+	panic(fmt.Errorf("invalid key %q in secret: %s/%s", keyFrom, s.Namespace, s.Name))
+}
+
+type testCase struct {
+	description string
+	test        func(ctx context.Context, c *fake.Clientset, provider *watchProvider)
+}
+
+type entry struct {
+	key   string
+	value string
+}
+
+type testCategory struct {
+	name    string
+	secret  *corev1.Secret
+	entries []entry
+}
+
+// Simpler re-implementation of the same method in the k8s library.
+func pollUntilContextTimeout(ctx context.Context, interval, timeout time.Duration, condition func(context.Context) (done bool, err error)) error {
```

Comment:
Prometheus upstream (2.43) did not have that method. It used k8s libraries which were too old. This is no longer an issue. Our 2.45 fork and upstream are both on v0.29 now so that method is available. I'll replace it with the built-in one. Thanks!